dags/operators/gcp_container_operator.py

```diff
@@ -0,0 +1,81 @@
+import os
+import subprocess
+import tempfile
+
+from airflow.contrib.operators.gcp_container_operator import GKEPodOperator
+
+
+KUBE_CONFIG_ENV_VAR = "KUBECONFIG"
+G_APP_CRED = "GOOGLE_APPLICATION_CREDENTIALS"
+
+
+class GKEPodOperator(GKEPodOperator):
+    """
+    We override execute and _set_env_from_extras methods to support:
+
+    - `CLOUDSDK_AUTH_CREDENTIAL_FILE_OVERRIDE` environment variable that is
+        set to the path of the Service Account JSON key file. This is neccesary
+        for gcloud to operate.
+
+    - Adjust when NamedTemporaryFile file descriptor is closed.
+
+    """
+    def execute(self, context):
+        # If gcp_conn_id is not specified gcloud will use the default
+        # service account credentials.
+        if self.gcp_conn_id:
+            from airflow.hooks.base_hook import BaseHook
+            # extras is a deserialized json object
+            extras = BaseHook.get_connection(self.gcp_conn_id).extra_dejson
+            # key_file only gets set if a json file is created from a JSON string in
+            # the web ui, else none
+            self._set_env_from_extras(extras=extras)
+
+        # Write config to a temp file and set the environment variable to point to it.
+        # This is to avoid race conditions of reading/writing a single file
+        with tempfile.NamedTemporaryFile() as conf_file:
+            os.environ[KUBE_CONFIG_ENV_VAR] = conf_file.name
+            # Attempt to get/update credentials
+            # We call gcloud directly instead of using google-cloud-python api
+            # because there is no way to write kubernetes config to a file, which is
+            # required by KubernetesPodOperator.
+            # The gcloud command looks at the env variable `KUBECONFIG` for where to save
+            # the kubernetes config file.
+            subprocess.check_call(
+                ["gcloud", "container", "clusters", "get-credentials",
+                 self.cluster_name,
+                 "--zone", self.location,
+                 "--project", self.project_id])
+
+            # Tell `KubernetesPodOperator` where the config file is located
+            self.config_file = os.environ[KUBE_CONFIG_ENV_VAR]
+            super(GKEPodOperator, self).execute(context)
+
+    def _set_env_from_extras(self, extras):
+        """
+        Sets the environment variable `GOOGLE_APPLICATION_CREDENTIALS` and
```

Comment:
I'm confused as to why CLOUDSDK_AUTH_CREDENTIAL_FILE_OVERRIDE works and GOOGLE_APPLICATION_CREDENTIALS does not.

Reply:
See `https://serverfault.com/questions/848580/how-to-use-google-application-credentials-with-gcloud-on-a-server`

Reply:
ah nice find